### DISCUSSION TITLE
Sticky Sessions for a Keycloak Cluster: Which guide is right?

### DISCUSSION THREAD

**Discussion by lukaseckert** [keycloak/keycloak#8692] [Q&A]
Hey everyone,
I'd like to ask if someone could clarify how to correctly setup sticky sessions for a Keycloak cluster behind an HAProxy load balancer for optimal performance. While there is a nice documentation at https://www.keycloak.org/docs/latest/server_installation/index.html#sticky-sessions, this (at least to my understanding ;) seems to contradict the guidelines in the Keycloak book by @stianst and @pedroigor 
 (Chapter 9 / Keeping session affinity).

**Variant A / keycloak.org: Stick to the node which is the Infinispan session owner**
From the workflow description at keycloak.org, HAProxy might route the initial request to node1. The Keycloak instance at node1 might create a session cookie like `AUTH_SESSION_ID=123.node2`, and from that point, HAProxy should route all requests to node2, because the session's Infinispan data is stored (only) on node2.
That means, HAProxy must parse the special `<session-id>.<owner-node-id>` syntax to read the ownerNodeId, so it can route the subsequent requests to the optimal node (here: node2). 
(So far I did not find a way to configure HAProxy accordingly. There is only a "prefix" but no "suffix" mode, so this results in AUTH_SESSION_IDs like `node1~123.node2` which seems to be the worst-case when all requests stick to node1 but the session data is on node2)

**Variant B / Keycloak Book: Stick to the node which got the initial request**
See https://books.google.de/books?id=WBsvEAAAQBAJ&pg=PA181&lpg=PA181
From the guide in the Keycloak book, the ownerId should be removed from the AUTH_SESSION_ID cookie by setting `shouldAttachRoute=false` on the `stickySessionEncoder`. At the same time, HAProxy stickyness should be configured independent from Keycloak by using a dedicated "KC_ROUTE" cookie, which is controlled by HAProxy only (`cookie KC_ROUTE insert indirect nocache`).

**Question:**
Variant B makes sense to me if the session data is always stored on the initially requested node, however that seems to contradict the Keycloak.org docs, which state that the session data is stored on one of the cluster nodes based on the hash of the sessionId. Any ideas on this?

Thanks for your support and the great work on Keycloak!

**Comment by stianst** [comment]
@mposolda @pedroigor

**Comment by 3XC1T3D** [comment]
I am also interested in a statement to this issue 👍

**Comment by pedroigor** [comment]
@lukaseckert Variant B is what makes more sense. We had a discussion a long time ago related to this but I can not find the thread right now.

But basically, doing Variant B is much more natural to most reverse proxies and also leverages their capabilities of choosing the best LB strategy. It also avoids misconfiguration and additional overhead when looking up entries because entries should be owned by the node that created that entry.

**Reply by lukaseckert** [reply]
Hi @pedroigor ,
thanks for your response!

**Comment by securite-apps** [comment]
![image](https://user-images.githubusercontent.com/90774712/146916567-cc7bcec4-b685-46f0-8af4-70b7f2a6f34c.png)

**Comment by r2coding** [comment]
What if you change the cache owners to meet the amount of instances you have? Then you should be able to answer the petition from any instance.

**Comment by MazizEsa** [comment]
Qurious whats the performance hit you got if you decided not to set the session affinity? Is it very significant?

**Reply by bjethwan** [reply]
`If the load balancer does not support or is not configured for [session stickiness](https://www.keycloak.org/docs/latest/server_installation/#sticky-sessions), the load over network in a cluster can increase significantly. The reason for this load is that each request that lands on a node that does not own the appropriate authentication session needs to retrieve and update the authentication session record in the owner node which involves a separate network transmission for both the retrieval and the storage.`
https://www.keycloak.org/docs/latest/server_admin/index.html#_limit-authentication-sessions

**Comment by slominskir** [comment]
1. Maybe the reverse proxy [guide](https://www.keycloak.org/server/reverseproxy) should include a new section on sticky sessions? 
2.  Some of the links in the discussion above no longer work such as https://www.keycloak.org/docs/latest/server_installation/index.html#sticky-sessions, is there an updated link?   
3. Does the new Quarkus build of Keycloak change the approach? 
4. Casually it would appear sticky sessions is purely a performance optimization, but my initial testing is suggesting some authentication flows (like Kerberos SPNEGO) may not work without it.

**Reply by lukaseckert** [reply]
For question 2: https://www.keycloak.org/docs/17.0/server_installation/index.html#sticky-sessions

**Reply by slominskir** [reply]
That updated link is for Keycloak 17.  Does the latest Keycloak docs have a section on sticky sessions?

**Comment by pedroigor** [comment]
@slominskir The approach is the same. We also include this [option](https://github.com/keycloak/keycloak/blob/59d9e3e3ee322198a3dcd811722cabe630a427e1/quarkus/dist/src/main/content/conf/keycloak.conf#L37), commented, in `conf/keycloak.conf` config file as a recommended option.

We can update the production guide to include:

```
If the load balancer does not support or is not configured for [session stickiness](https://www.keycloak.org/docs/latest/server_installation/#sticky-sessions), the load over network in a cluster can increase significantly. The reason for this load is that each request that lands on a node that does not own the appropriate authentication session needs to retrieve and update the authentication session record in the owner node which involves a separate network transmission for both the retrieval and the storage.
```

**Reply by slominskir** [reply]
Adding something to the production guide sounds good.  It appears the section on sticky sessions is missing in the latest docs, possibly because server installation is no longer a top level doc.  It isn't immediately obvious what [spi-sticky-session-encoder-infinispan-should-attach-route](https://www.keycloak.org/server/all-provider-config#_infinispan:~:text=spi%2Dsticky%2Dsession%2Dencoder%2Dinfinispan%2Dshould%2Dattach%2Droute) does exactly.

Bonus points if the docs actually include specific proxy examples (such as haproxy) or even better point to working docker compose examples.

**Reply by pedroigor** [reply]
@slominskir could you please check https://github.com/keycloak/keycloak/pull/16893?
